DOCSP-28006: Added conditional and set operations (#347)
mongodb/docs-java: source/fundamentals/aggregation/aggregation-expression-operations.txt

                                      then: "extreme temperature",
                                      else: "no advisory" } } } }

> Conditional Expression Operations

Reviewer: S: I believe we are omitting "Expression" in these titles. Also applies for set section below

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can perform conditional operations using the methods described in
this section.

.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - Java Method
     - Aggregation Pipeline Operator

   * - `cond() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlBoolean.html#cond(T,T)>`__
     - :manual:`$cond </reference/operator/aggregation/cond/>`

.. include:: /includes/fundamentals/agg-mql-diff.rst

> Suppose you want to find all documents where a customer purchased
> enough to qualify for a discount. The amount they need to spend for a discount
> varies if they are a team member.

Reviewer: S: rephrase for clarity and explaining the values used in the example

```suggestion
Suppose you want to find customers who spent enough money at a store to qualify for a discount. The amount they need to spend to qualify is $100 if they are a premium member and $200 if they are not.
```

> To decide whether the customer is a team member, check
> ``isTeamMember`` with the ``cond()`` method. To determine whether the sale
> was large enough, check the ``totalSale`` with the ``gte()`` method.

Reviewer: S: Rename team member to "premium member" to make things more clear

```suggestion
To check whether the customer is a premium member, check if
``isPremiumMember`` is ``true`` by using the ``cond()`` method. To determine whether the sale amount qualifies the customer for a discount, compare the ``totalSale`` field to the qualifying values by using the ``gte()`` method.
```

The following code shows the pipeline stage for this aggregation:

.. code-block:: java
   :copyable: true

>    Arrays.asList(match(
>             expr(current().getBoolean("isTeamMember").cond(
>                      current().getNumber("totalSale")
>                         .gte(of(100)),
>                      current().getNumber("totalSale")
>                         .gte(of(200))
>             ))
>    ));

Reviewer: I: format with breaks at each chained method

```suggestion
            Arrays.asList(match(
                    expr(current()
                            .getBoolean("isPremiumMember")
                            .cond(current()
                                            .getNumber("totalSale")
                                            .gte(of(100)),
                                    current()
                                            .getNumber("totalSale")
                                            .gte(of(200))))
            ));
```

The following code provides an equivalent aggregation pipeline in
the Query API:

.. code-block:: javascript
   :copyable: true

   [
     {
       $match: {
         $cond: {
            "$isTeamMember",
            { $gte: [ "$totalSale", 100 ] },
            { $gte: [ "$totalSale", 200 ] }
         }
       }
     }
   ]

Set Expression Operations
~~~~~~~~~~~~~~~~~~~~~~~~~

> Set expressions perform set operations on values of type ``MqlArray``, treating
> the arrays as sets.

Reviewer: Does this sound ok? Reworded to use "You can..." for consistency

```suggestion
You can perform set operations on values of type ``MqlArray``, as these operations treat
the arrays as sets.
```

Author: I broke it up into two sentences because it read better to me that way.

.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - Java Method
     - Aggregation Pipeline Operator

   * - `union() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlArray.html#union(com.mongodb.client.model.mql.MqlArray)>`__
     - :manual:`$setUnion </reference/operator/aggregation/setUnion/>`
       
.. include:: /includes/fundamentals/agg-mql-diff.rst

> Suppose you have a collection of music teacher information, including arrays of
> the students for each of their classes. You want to create a list of
> all the students for a teacher.

Reviewer: ```suggestion
Suppose you have a collection describing music teachers, and for each teacher, there are arrays of
the students for each of their classes. You want to create a list of all of the students that each teacher has by combining the rosters for their classes.
```

> To combine all of the ``pianoStudents`` and ``choirStudents``, use the

Reviewer: ```suggestion
To combine the ``pianoStudents`` and ``choirStudents`` arrays, use the
```

``union()`` method. To output the result as a new ``allStudents`` field, use
the ``project()`` method.

.. code-block:: java
   :copyable: true

>    Arrays.asList(project(fields(
>          computed("allStudents",
>                current().getArray("pianoStudents")
>                   .union((current().getArray("choirStudents"))))
>    )));

Reviewer: format

```suggestion
Arrays.asList(project(fields(
                    computed("allStudents",
                            current()
                                    .getArray("pianoStudents")
                                    .union((current()
                                            .getArray("choirStudents"))))
            )));
```

The following code provides an equivalent aggregation pipeline in
the Query API:

.. code-block:: javascript
   :copyable: true

   [
     {
       $project: {
         allStudents: {
            $setUnion: ["$pianoStudents", "$choirStudents"]

>          }

Author: I like the consistency on the page, so I think I am going to leave this as is.

                                      then: "extreme temperature",
                                      else: "no advisory" } } } }

Conditional Expression Operations

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Author: Resolving this since we discussed it in Slack.

Suppose you want to find all documents where a customer purchased
enough to qualify for a discount. The amount they need to spend for a discount
varies if they are a team member.

To decide whether the customer is a team member, check
``isTeamMember`` with the ``cond()`` method. To determine whether the sale
was large enough, check the ``totalSale`` with the ``gte()`` method.

The following code shows the pipeline stage for this aggregation:

.. code-block:: java
   :copyable: true

   Arrays.asList(match(
            expr(current().getBoolean("isTeamMember").cond(
                     current().getNumber("totalSale")
                        .gte(of(100)),
                     current().getNumber("totalSale")
                        .gte(of(200))
            ))
   ));

The following code provides an equivalent aggregation pipeline in
the Query API:

.. code-block:: javascript
   :copyable: true

   [
     {
       $match: {
         $cond: {
            "$isTeamMember",
            { $gte: [ "$totalSale", 100 ] },
            { $gte: [ "$totalSale", 200 ] }
         }
       }
     }
   ]

Set Expression Operations
~~~~~~~~~~~~~~~~~~~~~~~~~

> Set expressions perform set operations on values of type ``MqlArray``, treating
> the arrays as sets. Set expressions ignore the duplicate entries in each input
> array and the order of the elements.
> 
> If the set operation returns a set, the operation filters out duplicates in
> the result to output an array that only contains unique entries. The order of
> the elements in the output array is unspecified.
> 
> If a set contains a nested array element, the set expression does not descend
> into the nested array but evaluates the array at top-level.

Reviewer: ```suggestion
You can perform a set operation on values of type ``MqlArray``. Set operations treat arrays as sets.
```

The following table describes the set expression methods
available in the driver and corresponding expressions in the
Query API. The methods link to API documentation and the Query API

> expressions link to Server manual documentation.

Reviewer: S: Following the structure in #342, remove intro

Author: Understood. Will integrate this to other pull requests.

       | `switchDocumentOn() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlDocument.html#switchDocumentOn(java.util.function.Function)>`__
       | `switchIntegerOn() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlInteger.html#switchIntegerOn(java.util.function.Function)>`__
       | `switchMapOn() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlMap.html#switchMapOn(java.util.function.Function)>`__
       | `switchNumberOn() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#switchNumberOn(java.util.function.Function)>`__
       | `switchStringOn() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlString.html#switchStringOn(java.util.function.Function)>`__

     - :manual:`$switch </reference/operator/aggregation/switch/>`

.. include:: /includes/fundamentals/agg-mql-diff.rst

Suppose you have a collection which has a generic parameter used to store
miscellaneous customer data. What you want to do with that data varies,
depending on the type of data that is stored.

To change what you are doing based on the data type of ``miscData``, use the

> ``switchOn()`` method. If the data type is ``String``, just copy the data

Reviewer 2: - `String` is an actual type in Java. The closest corresponding type in our BSON API is `BsonString` (which is what the actual check will be against), and in this MQL API, it's `MqlString`. To avoid confusion, it seems best to use the form "if the value is a string...".
- It seems odd to invoke the "copied" concept here - maybe "use/take the data as-is" or something.
-

without altering it. If the data type is ``Boolean``, check the value with the
``cond()`` method and return a string indicating whether the customer is a
premium member. If the data type is ``Array``, get the first element by using
the ``elementAt()`` method. If it is not one of these types, record a message
by using ``defaults()``. Regardless of the type, use the ``project()`` method
to output the result as ``result``.

The following code shows the pipeline stage for this aggregation:

.. code-block:: java
   :copyable: true

   Arrays.asList(project(fields(
         computed("result",
               current().getField("miscData").switchOn(field -> field
                        .isString(s -> s)
                        .isBoolean(b -> b.cond(
                                 of("Customer is premium member"),
                                 of("Customer is not premium member")))

>                         .isArray(a -> a.elementAt(0).asString())

Reviewer 2: ```suggestion
                        .<MqlString>isArray(a -> a.first())
```
This asserts the type of the array, since you expect it will be a string.

>          computed("result",
>                current().getField("miscData").switchOn(field -> field
>                         .isString(s -> s)
>                         .isBoolean(b -> b.cond(
>                                  of("Customer is premium member"),
>                                  of("Customer is not premium member")))
>                         .isArray(a -> a.elementAt(0).asString())
>                         .defaults(d -> of("Not an anticipated type"))
>                ))

Reviewer 2: There are two "patterns" where switch applies: 1) you have the same data, stored in different formats, and 2) you want to branch off on specific conditions (much like a typical switch, or a sequence of if-else statements).

This example is good in that it covers many important points for number 1, but I found it difficult to follow, I think because it seemed arbitrary or artificial. 

I think it could be improved by explaining (or implying / assuming) that, in the example, there was a "member" field that contained true or false indicating whether the customer had a premium membership, but then perhaps other membership levels were introduced (like gold and platinum), which were stored as strings, and evidently customers can now also have multiple memberships, the first (perhaps `.last`?) of which is their most recent.

```suggestion
         computed("membershipLevel",
               current().getField("member").switchOn(field -> field
                        .isString(s -> s)
                        .isBoolean(b -> b.cond(of("premium"), of("regular")))
                        .<MqlString>isArray(a -> a.last())
                        .defaults(d -> of("regular"))
               ))
```

   )));

The following code provides an equivalent aggregation pipeline in
the Query API:

.. code-block:: javascript
   :copyable: true

   [
     {
       $project: {
         result: {
           $switch: {
             branches: [
               { case:
                 { $eq: [ { $type: "$miscData" }, "string" ] },
                 then:
                   "$miscData" },
               { case:
                 { $eq: [ { $type: "$miscData" }, "bool" ] },
                 then:
                   { $cond: { if: "$miscData",
                     then: "Customer is premium member",
                     else: "Customer is not premium member" } } },
               { case:
                 { $eq: [ { $type: "$miscData" }, "array" ] },
                 then:

>                    { $toString: { $arrayElemAt: [ "$miscData", 0 ] } } }

Reviewer 2: Is the JSON condensed/reformatted here?

             ],
             default: "Not an anticipated type"
           }
         }
       }
     }
   ]

Set Operations
~~~~~~~~~~~~~~

You can perform set operations on values of type ``MqlArray``. These

> operations treat the arrays as sets.

Author: This is something I'd like to record and push back until we are all merged together and can reorganize them all under headings as we discussed in Slack.

.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - Java Method
     - Aggregation Pipeline Operator

   * - `union() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlArray.html#union(com.mongodb.client.model.mql.MqlArray)>`__

>      - :manual:`$setUnion </reference/operator/aggregation/setUnion/>`

Author: It's in the array operations (this is part of why I want to do the re-org and final verification of methods once we have a full topic with all these sections merged together).
